pkg/gadgets/interface_test.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2019-2021 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gadgets
+
+import (
+	"testing"
+)
+
+type fakeTraceFactory struct {
+	BaseFactory
+
+	deletedHasBeenCalled bool
+}
+
+type fakeTrace struct {
+}
+
+func newFactory() TraceFactory {
+	f := &fakeTraceFactory{}
+	f.DeleteTrace = f.deleteTrace
+	return f
+}
+
+func (f *fakeTraceFactory) deleteTrace(name string, t interface{}) {
+	f.deletedHasBeenCalled = true
+}
+
+func TestDeleteTrace(t *testing.T) {
```

Comment:
Mmmm I think this test is already covered in the controller tests:
https://github.com/kinvolk/inspektor-gadget/blob/f1ac14ede7a6b22fbfd879636e402c24619cacd9/pkg/controllers/trace_controller_test.go#L228 

But the `FakeFactory`'s deletion is not following the same pattern we used in the actual gadgets and it prevented us from discovering the issue reported by this PR before, i.e. the delete function will not be called. 

In fact, if we change the `FakeFactory`'s deletion from ...
```
func (f *FakeFactory) Delete(name string)
```
... to ...
```
func (f *TraceFactory) DeleteTrace(name string, trace interface{})
```
... , as the actual gadgets, the test will start failing.

The confirmation is that after applying this PR and setting the delete function as requested now, the issue is also resolved there:
```
        fakeFactory := &FakeFactory{}
	fakeFactory.calls = make(map[string]struct{})
	fakeFactory.DeleteTrace = fakeFactory.deleteTrace
```

If my understanding is correct, I would prefer to fix the controller tests and avoid adding this new separated test. What do you think?

cc @mauriciovasquezbernal

Reply:
Well spotted for the controller tests. Yes, the controller tests should be updated to use the same pattern.

I am not sure about removing the new separated test because, as opposed to controller tests, it can be run simply with `go test` without downloading extra dependencies (kube-apiserver etcd kubectl).

Reply:
Branch updated.

Reply:
> as opposed to controller tests, it can be run simply with go test without downloading extra dependencies (kube-apiserver etcd kubectl).

That's true. It took me a while to understand how to run the controller tests 😅 (BTW, we should document it). 

The point is that we would have two different tests to verify the same operation ... Mmm it's not a big deal, I just don't see the need.